### PR TITLE
Extract and execute JUnit5 tests

### PR DESCRIPTION
**Pull request by luoos**

### CODE REVIEW

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala [hunk 1]
```diff
@@ -20,7 +25,103 @@ object TestFramework {
...
+        } catch {
+            case e: NoClassDefFoundError => Option.empty
+        }
+    }
+
+    override def toString(): String = "JUnit"
+
+    // the splitor to split the class name and method name
+    // for the full qualified name of JUnit 4 test
+    // like com.package.JUnit4TestClass.TestA
```
Comment (luoos) [inline testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:90]: This is the original behavior for JUnit 4.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala [hunk 2]
```diff
@@ -20,7 +25,103 @@ object TestFramework {
...
+        } catch {
+            case e: NoClassDefFoundError => Option.empty
+        }
+    }
+
+    override def toString(): String = "JUnit5"
+
+    // the splitor to split the class name and method name
+    // for the full qualified name of JUnit 5 test
+    // like com.package.JUnit5TestClass#TestA
```
Comment (luoos) [inline testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:125]: This is the correct way to write the fully qualified name for Junit 5

File: pom.xml [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
     <groupId>edu.illinois.cs</groupId>
     <artifactId>testrunner</artifactId>
-    <version>1.1-SNAPSHOT</version>
+    <version>1.2-SNAPSHOT</version>
```
Comment (august782) [inline pom.xml:8]: I can see why it would be good to update version number for big change like adding JUnit 5, but we have yet to release 1.1 (without SNAPSHOT). We should either keep this as 1.1-SNAPSHOT and release this as part of 1.1, or release 1.1 before these changes and start 1.2-SNAPSHOT with this JUnit 5 change.

File: testrunner-running/pom.xml [hunk 4]
```diff
@@ -41,6 +41,11 @@
             <groupId>xmlunit</groupId>
             <artifactId>xmlunit</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.junit.platform</groupId>
+            <artifactId>junit-platform-launcher</artifactId>
+            <version>1.1.0</version>
```
Comment (august782) [inline testrunner-running/pom.xml:47]: Since this dependency is included in parent's dependencyManagement, there is no need to include version here.

File: pom.xml (hunk 3 above)
Comment (luoos) [reply pom.xml]: Oh! I thought 1.1 have been released. And Done.

File: testrunner-running/pom.xml (hunk 4 above)
Comment (luoos) [reply testrunner-running/pom.xml]: Done

File: testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala [hunk 5]
```diff
@@ -3,6 +3,8 @@ package edu.illinois.cs.testrunner.mavenplugin
 import java.nio.file.{Files, Paths}
 
 import edu.illinois.cs.testrunner.configuration.Configuration
+import edu.illinois.cs.testrunner.data.framework.TestFramework
```
Comment (winglam) [inline testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala]: Do we actually need this import?

File: testrunner-running/pom.xml [hunk 6]
```diff
@@ -41,6 +41,10 @@
             <groupId>xmlunit</groupId>
             <artifactId>xmlunit</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.junit.platform</groupId>
+            <artifactId>junit-platform-launcher</artifactId>
```
Comment (winglam) [inline testrunner-running/pom.xml:46]: Do you need this dependency declaration if you already have it in the parent pom?

File: testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala (hunk 5 above)
Comment (august782) [inline testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala:6]: Why is the TestFramework import necessary?

Comment (luoos) [reply testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala]: Oh, yea. It's redundant. Deleted.

Comment (luoos) [reply testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala]: Deleted.

File: testrunner-running/pom.xml (hunk 6 above)
Comment (luoos) [reply testrunner-running/pom.xml:46]: Yes, we need it. Otherwise, it fails to compile.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
+import java.util.List;
 import scala.collection.JavaConverters._
-
-sealed abstract class TestFramework
-case object JUnit extends TestFramework
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
 
 object TestFramework {
     def testFramework(project: MavenProject): Option[TestFramework] = {
```
Comment (august782) [inline testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: It seems like this method now returns the relevant TestFramework as long as there is just one, but right now it is only relevant for JUnit 4. Should it be modified to also check if junit-jupiter is there and return JUnit5?

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.platform.engine.discovery.MethodSelector;
+import org.junit.platform.launcher.Launcher;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.launcher.core.LauncherFactory;
+
+// Use JUnit 5 API to run tests, including JUnit 4 and 5 tests.
+public class JUnit5TestExecutor {
+
+    public static TestRunResult runTestsSeparately(
```
Comment (winglam) [inline testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: We discussed an improvement that seems to be missing from this PR (but can be added later): when running tests separately, consecutive tests from the same test class would still be run together (ClassA.t1, ClassA.t2, ClassB.t3, ClassA.t4 would only invoke the launcher 3 times and not 4 times) 

Where are these potential improvements being tracked at the moment?

File: testrunner-running/pom.xml (hunk 6 above)
Comment (winglam) [reply testrunner-running/pom.xml:46]: Do you need it on the parent pom as well then?

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala [hunk 9]
```diff
@@ -29,6 +29,9 @@ object Executor {
             case "JUnit" =>
                 JUnitTestExecutor.runOrder(testRunId, tests, true, false)
                     .writeTo(outputPath.toAbsolutePath.toString)
+            case "JUnit5" =>
+                JUnit5TestExecutor.runTestsSeparately(testRunId, tests)
+                                  .writeTo(outputPath.toAbsolutePath.toString)
```
Comment (august782) [inline testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala:34]: Shift the whitespace here to match how it is for line 31.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Yea, I will add a TODO here.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala (hunk 9 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala:34]: I think line 34 looks better than line 31. How about I change line 31 to match line 34?

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (august782) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Why is the method called runTestsSeparately while in JUnitTestExecutor the corresponding method seems to be runOrder? Is there something special about JUnit 5 such that it needs to be called runTestsSeparately?

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala (hunk 9 above)
Comment (august782) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala:34]: That's fine, we do want consistency here. We should eventually implement some checkstyle for this codebase as well.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: There was a long conversation about this. In short, I haven't found a way to execute tests with full control of the order, if we don't run "separately". This is a workaround so that we have full control of the execution order. Refer [here](https://github.com/luoos/junit5-testrunner-demo/blob/1d0e4388b1193902eac4b227c37c12fc0e31eae9/src/main/java/com/luojl/demo/TestRunner.java#L17) to some details.

File: testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala [hunk 10]
```diff
@@ -17,14 +18,14 @@ import scala.collection.JavaConverters._
 class TestRunner extends TestPlugin {
     def tests(source: String, project: MavenProject): Stream[String] = {
         if (source == null) {
-            TestLocator.tests(project)
+            TestLocator.tests(project, JUnit)
```
Comment (august782) [inline testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala:21]: Based on these changes, is it intentional that this TestRunner only applies to JUnit 4 tests? Should it be changed to handle JUnit 5 as well?

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (winglam) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: IIRC the bigger problem we have wouldn't exactly be how these JUnit 5 tests are being run (i.e., everytime a test is run, it's @BeforeClass and @AfterClass would run), but that this behavior would be inconsistent with what we are doing for JUnit 4 tests (i.e., a test class's @BeforeClass is only run once, the very first time a test method from the class is run even if tests from other classes interleave).

As I don't see strong reasons to choose one behavior over the other, we should just look to include making the way JUnit 4 and JUnit 5 treat @BeforeClass/@AfterClass consistent in our list of TODOs/Caveats.

File: testrunner-running/pom.xml (hunk 6 above)
Comment (august782) [reply testrunner-running/pom.xml:46]: Since the parent does not have any code, it doesn't actually need any real dependency, but having them declared under dependencyManagement there is fine as to keep consistent versions throughout.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala (hunk 7 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: I agree that this method doesn't look right. This method has only one output `TestFramework`, but there are two possible frameworks.

If the maven project has two frameworks (J4 and J5), what is the expected output of this method? It seems that my newly added method `getListOfFrameworks` can cover this method, which means that we don't need this method anymore.

File: testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala (hunk 10 above)
Comment (luoos) [reply testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala:21]: It seems a legacy code. Based on the current usage of iDFlakies, this class (TestRunner) will not be used. The real class being used is DetectorPlugin.

I change this line to pass compile because I modify this signature of TestLocator.tests.

Not very sure what to do with this class.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala (hunk 9 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala:34]: Done

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Done. Add some annotation here.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala (hunk 9 above)
Comment (winglam) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala:34]: It looks like you are off by one space still on either line 31 or 34.

Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/Executor.scala:34]: I think the rule is to match the dot position with the above line.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (winglam) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: I don't understand what you want a reader of the TODO to do. The problem as I said is mainly that the behavior for JUnit 4 and JUnit 5 is different in testrunner now.

The current way that JUnit 5 tests are run doesn't seem problematic to me. If it is hard (or impossible) to replicate JUnit 4's behavior in JUnit 5, shouldn't this TODO just tell the reader to implement JUnit 5's behavior for JUnit 4?

The text you have now largely explains why it is difficult to implement JUnit 4's behavior in JUnit 5, which is fine to keep. However, I don't know why "optimal solution" is to replicate JUnit 4's behavior in JUnit 5. That to me, seems to be a solution, but as you pointed out is hard or impossible of a solution anyway.

File: testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala (hunk 10 above)
Comment (august782) [reply testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala:21]: This seems to be a sample testrunner that simply runs the tests it can find within the project. Given the changes, right now it only runs JUnit 4 tests and does not run any JUnit 5 tests. To support JUnit 5, maybe add some code that tries both JUnit and JUnit5 as the parameter, so if it fails to get tests using one TestFramework it will try the other.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (winglam) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: I also don't see any mentioning of the TODO to add the feature of "when running tests separately, consecutive tests from the same test class would still be run together and ``@BeforeClass` and `@AfterClass` would only be run once for them".

Overall, I see two TODOs when we merge your current changes:
(1) the feature of "when running tests separately, consecutive tests from the same test class would still be run together and ``@BeforeClass` and `@AfterClass` would only be run once for them"
(2) make JUnit 4 and JUnit 5's behavior consistent in how it treats `@BeforeClass` and `@AfterClass` (you can then clarify for this TODO how one of the solutions (making JUnit 5 be like JUnit 4) is hard or impossible

Can you please clarify the two TODOs here?

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala (hunk 7 above)
Comment (august782) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: So from what I understand, the TestFramework is used to determine what runner or test executor to use to run tests. The TestFramework is also used to determine what are the tests to run, which is based off of annotations. Is it possible for JUnit 5 to run JUnit 4 style tests? If a project actually does have both JUnit 4 and JUnit 5 tests, does it utilize JUnit 5 to actually run the JUnit 4 tests (i.e., the dependency on JUnit 4 is just for compilation purposes)?

Basically, if JUnit 5 can run JUnit 4 tests, then perhaps the code for JUnit 5 could be updated to also service JUnit 4 tests in case both are present?

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Let's step back a little bit, and think about a more fundamental question. The reason for the "separate" method is to get **full** control of the test order. But do we really need that?

The JUnit 5 API already provides a simple API to get the **almost-full** control of the test order. There are just two exceptions: 1) interleaved class 2) Order annotation.

However, the above two situations will never happen in the real case.

1) **Interleaved class**: When a user runs a test suite, it's impossible to run some tests in ClassA, then some in ClassB, then again some in ClassA. Even if iDFlakies can run tests in a class-interleaved way, the "flaky" tests found in this way are just a false alert.
2) **Order annotation**: Users already specify the order for their tests. If we force to change the order and find some "flaky" tests, that would be a false alert as well.

All in all, my point is that we don't need the "separate" method to get the **full** control. Using the simple JUnit 5 API (like [this](https://github.com/luoos/junit5-testrunner-demo/blob/1d0e4388b1193902eac4b227c37c12fc0e31eae9/src/main/java/com/luojl/demo/TestRunner.java#L15)) to get the "**almost-full**" control is good enough. In this way, there is no performance overhead and @BeforeClass-things will behave like JUnit 4.

Based on the above behavior, we need to modify the shuffling logic: we need to make sure 1) there is no class-interleaved case, and 2) don't shuffle tests with Order annotation.

@winglam what do you think?

Comment (winglam) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Are you suggesting that instead of merging `runTestsSeparately` you would rather just merge an implementation that grants you almost-full control of the test order? If so, why not strive to have both at some point and you can just merge `runTestsSeparately` now? (IIRC you already had both implementations ready the last time we met and Darko's suggestion was just to add a configuration option to choose between the two.)

Your concern with `runTestsSeparately` is largely centered around the fact that it creates "false alerts". Indeed, if one is using `runTestsSeparately` to detect OD tests, then the detected tests may be construed as false positives to the developers. However, imagine iFixFlakies trying to find a cleaner. In such a case, it doesn't matter that the cleaner requires the interleaving of two classes or violates the order annotation.

As I have been saying, there isn't a strong reason that one implementation is strictly better than another as each implementation will benefit one use case or another. The most problematic part of your changes here is simply that the current implementation of testrunner will handle JUnit 4 and JUnit 5 tests differently (i.e., JUnit 4 runs `@BeforeClass` once for every test class no matter the order, while JUnit5 may run `@BeforeClass` many times). For now, it suffices to document this problem as a TODO so that when you or others have more time, it can be addressed then.

File: testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala (hunk 10 above)
Comment (luoos) [reply testrunner-maven-plugin/src/main/scala/edu/illinois/cs/testrunner/mavenplugin/TestRunner.scala]: Done. Please resolve this conversation if it looks good to you.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java (hunk 8 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Sounds reasonable with the point related to iFixFlakies. OK, let's try to merge this "separate" one first. I will modify the TODO to highlight the different behavior between J4 and J5.

Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Updated the annotation. Please resolve this conversation if it looks good to you.

Comment (winglam) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Thanks for adding the detailed information for the TODO related to JUnit 4 and JUnit 5 differences.

Can you please add a TODO for the following with the relevant details as well?
> the feature of "when running tests separately, consecutive tests from the same test class would still be run together and `@BeforeClass` and `@AfterClass` would only be run once for them"

Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/execution/JUnit5TestExecutor.java:16]: Updated. But I think the effect of that improvement would be limited, especially when there are thousands of tests. In that case, the possibility to find consecutive and same-class tests is pretty low.

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala (hunk 7 above)
Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: It's possible for JUnit 5 framework to run JUnit 4 style tests, and I have an [implementation](https://github.com/luoos/testrunner/commit/6d21eb7f0979b2c2f50d278a7b1a3672d1e57c8e) for it. But there is an issue. Using J5 framework to run J4 tests needs a special runtime dependency - [junit-vintage-engine](https://junit.org/junit5/docs/current/api/org.junit.vintage.engine/module-summary.html). This issue is pretty tricky. And I discussed it with the professor.

In short, when the testrunner starts a new process to run tests, we need to add the jar file path of the `junit-vintage-engine` dependency to `classpath` (-cp) parameters. However, it's a bit tricky to do that. One simple way is to modify the target project's pom.xml, adding the `junit-vintage-engine` as a dependency. But this method is verbose and error-prone.

Combining that we don't have full control of test order with J5 API (if not use the "separate" method as a workaround), it would be safer, for now, to keep using the original J4 executor to run J4 tests, and the J5 executor to run J5 tests.

Comment (august782) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: I assume you mean when there are a mix of JUnit 4 and 5 tests then you would need the junit-vintage-engine dependency to run them (because if there was only JUnit 4 tests and no dependency on junit-jupiter, then you wouldn't need to have JUnit 5 run the JUnit 4 tests). However, if there is a mix, does the target project need to have junit-vintage-engine as a dependency already to even run those tests? Should that dependency already be there in such a case?

Concerning using JUnit 5 API to run JUnit 4 tests, is it that running them in alphabetical order a built-in requirement? In other words, if one would to normally run JUnit 4 tests through JUnit 5 in a project, does the specification for JUnit 5 say that those tests would be run in alphabetical order?

Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: For a mixed project, it typically has 1) `junit` dependency to run J4 tests (*J4 framework to J4 tests*), 2) `junit-jupiter` dependency to run J5 tests (*J5 framework to run J5 tests*). It doesn't need `junit-vintage-engine`. But what we're trying to do here, is to use J5 API to run J4 tests, that's quite different. So we need to add the `junit-vintage-engine` to the target project.

"running them in alphabetical order" is what I observed with my [demo project](https://github.com/luoos/junit5-testrunner-demo), I think that's built-in behavior.

**Review by winglam — APPROVED** [review]
Overall, the changes LGTM. The remaining issues can be improved later. @august782 feel free to merge once you are content as well

File: testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala (hunk 7 above)
Comment (august782) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: How about we change this method to check for which dependency exists and returns the one that does exist, and if both exist then either throw some exception about there being multiple relevant dependencies or returning Option.empty. We should leave a note about this issue that this method assumes only no mix of dependencies and one should use getListOfFrameworks for the more general case. We can look into this more later if we end up needing more support for this mix.

From what I understand, while one can run both JUnit 4 and JUnit 5 tests by having both dependencies, the "correct", recommended way is to include that junit-vintage-engine so JUnit 5 can run JUnit 4 tests. I suppose we need to support even this other way with both dependencies for the projects that don't play so nice, but I don't think it's necessarily wrong to always assume junit-vintage-engine is not there.

Comment (luoos) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: Updated the testFramework method as required.

I'm not sure about the "correct" practice for a J4 and J5 mixed project. But for J4-only project, it definitely doesn't contain `junit-vintage-engine` dependency. So if we want to use J5 API to run J4 tests, we still need to add that dependency. So we can't simply assume that dependency exists.

Comment (august782) [reply testrunner-running/src/main/scala/edu/illinois/cs/testrunner/data/framework/TestFramework.scala:16]: Thanks for making the change. This should be good for now to merge.

To elaborate a bit more, yes, if the project only has JUnit 4 (where it doesn't even have any dependency on junit-jupiter), then we should for sure run using the JUnit 4 API. I am not suggesting that we only use JUnit 5 to run tests, regardless of present dependencies. If the project has both JUnit 4 and JUnit 5 tests, then I think it is recommended that the project use junit-vintage-engine to run the JUnit 4 style tests.
